Add PeerNotRespondingError

manoranjith opened the pull request:
This error identifies the scenario where the context expires when waiting for a response for a channel update or a channel proposal from the peer.

This was the best place I was able to detect this case. But I am not sure if context expiring could always be treated as "Peer Failing To Respond". Request your comments on this.

ggwpez commented on client/update.go:
```diff
@@ -411,3 +419,8 @@ func makeChannelUpdate(next *channel.State, actor channel.Index) ChannelUpdate {
 }
+
+// Error implements error interface for PeerNotRespondingError.
+func (e PeerNotRespondingError) Error() string {
```
Could you add constructors for both error types which also accept an error and use the `WithStack` function?  
Then the context error would be passed on instead of being discarded (like now). You can take a look at `channel/errors.go`.

manoranjith replied:
I have added `newRequestTimedOutError` function.

ggwpez commented on client/update.go:
```diff
@@ -411,3 +419,8 @@ func makeChannelUpdate(next *channel.State, actor channel.Index) ChannelUpdate {
 		ActorIdx: actor,
 	}
 }
+
```
I still think it is sad that we now mix error checking functions like `Is…Error` and `errors.Is` but OK… its too late now.

manoranjith replied:
I have used `errors.As` in the tests for the newly added error. Should I instead add an `Is..Error` function that internal uses `errors.Cause` as in other places in the project ? @ggwpez

matthiasgeihs replied:
It would be nice if it was consistent across the framework. On the other hand, I don't like introducing Is...Error functions all over the place, as this is pretty much redundant to the errors.Is functionality. Or am I wrong on this? Any other views and suggestions on this are appreciated.

manoranjith replied:
I agree with you @matthiasgeihs. Personally I also prefer to use `errors.Is` over `Is..Error` function.

ggwpez replied:
Ok then lets leave it as is.

ggwpez commented on client/update.go:
```diff
@@ -185,6 +189,10 @@ func (c *Channel) update(ctx context.Context, next *channel.State) (err error) {
 	if err != nil {
+		// Context expiry here implies we have timedout waiting for response from the peer.
+		if pcontext.IsContextError(err) {
+			return errors.WithStack(PeerNotRespondingError("peer did not respond to the channel proposal"))
```
This function still uses `err` in a defer in line 161, better set it to the newly created error here.

ggwpez commented on client/update.go:
```diff
@@ -185,6 +189,10 @@ func (c *Channel) update(ctx context.Context, next *channel.State) (err error) {
 
 	pidx, res, err := resRecv.Next(ctx)
 	if err != nil {
+		// Context expiry here implies we have timedout waiting for response from the peer.
```
Please add a comment to `Update` in L133 about the new error.

ggwpez replied:
```suggestion
```

ggwpez commented on client/proposal.go:
```diff
@@ -264,6 +265,10 @@ func (c *Client) proposeTwoPartyChannel(
 
 	env, err := receiver.Next(ctx)
 	if err != nil {
+		// Context expiry here implies we have timedout waiting for response from the peer.
```
Pleas add a comment to `ProposeChannel` in L132 about the error.

ggwpez replied:
```suggestion
```
I think the code is easy enough to understand here.

ggwpez commented on client/update.go:
```diff
@@ -93,6 +94,9 @@ type (
 	}
+
+	// PeerNotRespondingError indicates that a peer has not responded within the expected time period.
+	PeerNotRespondingError string
```
Im not quite happy with the name, since it sound like the peer is not responding at all.  
But in this case it only did not respond to a specific request (or we did not receive it).  
Maybe something like `RequestTimedOutError` or `NoResponseError`.

manoranjith replied:
I think `RequestTimedOutError` looks better. I will use this one.

ggwpez reviewed (changes requested):
> But I am not sure if context expiring could always be treated as "Peer Failing To Respond".

When the user closes the Channel, Client or Context while waiting for a request this error would also be triggered, but i think this is ok.  
  
Regarding testing; this is untested even before you modified it :man_facepalming: Maybe i find a place to test it.

manoranjith commented:
@ggwpez I have addresses most of the review comments and also rebased onto the tip of dev branch. I have left the two unaddressed comments in the unresolved state.

matthiasgeihs commented on client/proposal.go:
```diff
@@ -264,6 +266,9 @@ func (c *Client) proposeTwoPartyChannel(
 	env, err := receiver.Next(ctx)
 	if err != nil {
+		if pcontext.IsContextError(err) {
+			return nil, newRequestTimedOutError("channel proposal", err.Error())
```
Wouldn't it be better to wrap the error object instead of just the error string?

manoranjith replied:
I wrap the `newRequestTimedOut` error instead of the actual error because it is this error that the user of the API will be checking for. If we do it otherwise: wrap the actual error and add the `newRequestTimedOutError` as a message, I think it would not be possible for to check for `RequestTimedOutError`.

manoranjith replied:
I add the error string, so that the details of the original error message is still retained.

manoranjith replied:
@matthiasgeihs However, if the underlying error is to be retained along with this error. Then one option is to store the actual error in the `RequestTimeOutError` struct and write an `Unwrap` method for it (see [here](https://play.golang.org/p/iJKrTEQApuq)) which will return the underlying error.

matthiasgeihs reviewed (commented):
Please see my comment on the custom error object. I was surprised that it does not wrap the underlying error, but just the error string.

ggwpez commented on client/update_test.go:
```diff
@@ -42,3 +43,19 @@ func TestChannel_Update_NilArgs(t *testing.T) {
 }
+
+func TestRequestTimedOutError(t *testing.T) {
+	var err error = errors.WithStack(newRequestTimedOutError("", ""))
```
It should already have a stacktrace now.

manoranjith replied:
Yeah, i missed to remove `WithStack` after adding the `newRequestTimedOutError` in previous review iteration. It have updated it.

ggwpez commented on client/update.go:
```diff
@@ -185,6 +191,10 @@ func (c *Channel) update(ctx context.Context, next *channel.State) (err error) {
 	pidx, res, err := resRecv.Next(ctx)
 	if err != nil {
+		if pcontext.IsContextError(err) {
+			err := newRequestTimedOutError("channel update", err.Error())
```
```suggestion
			err = newRequestTimedOutError("channel update", err.Error())
```

manoranjith replied:
Updated.

ggwpez commented on client/update.go:
```diff
@@ -411,3 +422,12 @@ func makeChannelUpdate(next *channel.State, actor channel.Index) ChannelUpdate {
 	}
 }
+
+// Error implements error interface for PeerNotRespondingError.
```
```suggestion
// Error implements error interface for RequestTimedOutError
.
```

manoranjith replied:
Updated.

manoranjith commented:
I have addressed all your review comments (3) in the 2nd iteration. @ggwpez

ggwpez reviewed (approved):
Im fine with this.
